src/main/java/slimeknights/tconstruct/smeltery/block/entity/CastingTankBlockEntity.java

```diff
@@ -0,0 +1,367 @@
+    this(TinkerSmeltery.castingTank.get(), pos, state, block);
+  }
+
+  /** Extendable constructor */
+  @SuppressWarnings("WeakerAccess")
+  protected CastingTankBlockEntity(BlockEntityType<?> type, BlockPos pos, BlockState state, ITankBlock block) {
+    super(type, pos, state, NAME, 2, 1);
+    tank = new FluidTankAnimated(block.getCapacity(), this);
+    fluidHolder = LazyOptional.of(() -> tank);
+    itemHandler = new SidedInvWrapper(this, Direction.DOWN);
+  }
+
+  /**
+   * Called from {@link slimeknights.tconstruct.smeltery.block.CastingTankBlock#use(BlockState, Level, BlockPos, Player, InteractionHand, BlockHitResult)}
+   * @param player Player activating the block.
+   */
+  public void interact(Player player, InteractionHand hand, BlockHitResult hit) {
+    // skip client side
+    if (level == null || level.isClientSide) {
+      return;
+    }
+
+    ItemStack input = getItem(INPUT);
+    ItemStack output = getItem(OUTPUT);
+    ItemStack heldCopy = player.getItemInHand(hand).copy();
+
+    // if there is an item in the output slot, take it
+    if (!output.isEmpty()) {
+      setItem(OUTPUT, ItemStack.EMPTY);
+      ItemHandlerHelper.giveItemToPlayer(player, output, player.getInventory().selected);
+    // otherwise, interact with the tank
+    } else if (!FluidTransferHelper.interactWithTank(level, worldPosition, player, hand, hit)
+      // even if we did "interact" with the tank, still need to check whether the item actually changed
+      || ItemStack.isSameItemSameTags(heldCopy, player.getItemInHand(hand))) {
+
+      // if the tank wasn't interacted with, first try to place a held item in the input
+      ItemStack held = player.getItemInHand(hand);
+      if (canPlaceItem(INPUT, held)) {
+        setItem(INPUT, held.split(1));
+        // otherwise, try to take the item from the input slot
+      } else if (!input.isEmpty()) {
+        setItem(INPUT, ItemStack.EMPTY);
+        ItemHandlerHelper.giveItemToPlayer(player, input, player.getInventory().selected);
+      }
+    }
+  }
+
+  @Override
+  public void setItem(int slot, ItemStack itemstack) {
+    ItemStack currentInput = getItem(INPUT);
+    super.setItem(slot, itemstack);
+    // if we are setting the input item and it actually changed, try to process it
+    if (slot == INPUT && !ItemStack.matches(currentInput, itemstack)) {
+      tryToProcessItem();
+    }
+  }
+
+  public void setInputItemWithoutTryingToProcess(ItemStack stack) {
+    super.setItem(INPUT, stack);
+  }
+
+  @Override
+  public boolean canPlaceItem(int pIndex, ItemStack pStack) {
+    if (pIndex == INPUT) {
+      return getItem(INPUT).isEmpty() && getItem(OUTPUT).isEmpty() && !pStack.isEmpty() && (
+        // check the various options for some sort of fluid-containing stack
+        FluidContainerTransferManager.INSTANCE.mayHaveTransfer(pStack)
+          || pStack.getCapability(ForgeCapabilities.FLUID_HANDLER_ITEM).isPresent()
+          || pStack.getItem() instanceof BucketItem
```

Comment:
Nit pick, but the bucket item check is probably faster than the capability check as forge capabilities don't use a map data structure; maybe in 1.21 the opposite order is better.